### PR TITLE
REALMC-11535: add destructuring, rest, spread, default functions

### PR DESCRIPTION
**Pull request by Gabri3l**
I'm trying to take baby steps here and port fewer changes at a time that are stable with all of our owns tests passing.

Also:

- Avoid calling iterator's return method again when it throws 
- Fixed dynamic variable resolution when a parent lexical binding exists
- Support dereferencing in ExportTo().
-  Make sure variables dynamically bound in parameter scope conflict with parameter bindings.
- Fixed setting the stack pointer in enterFuncBody.
-  Support for assignment patterns in for-in and for-of loops. Support for spread elements in call arguments. Fixed various issues related to destructuring assignments.

These are cherry picked from later on so that we can be in stable state without breaking any of our tests:
- Fixed binding order for global function declarations when there are duplicate names
- Fixed argument variable reference resolution in stashless functions

[Evergreen Patch with these changes](https://spruce.mongodb.com/version/61f44526d1fe070f24f47b8d/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) ✅

### CODE REVIEW

File: destruct.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+func (d *destructKeyedSource) _putProp(name unistring.String, value Value, writable, enumerable, configurable bool) Value {
+	return d.w()._putProp(name, value, writable, enumerable, configurable)
+}
+
+func (d *destructKeyedSource) _putSym(s *Symbol, prop Value) {
+	d.w()._putSym(s, prop)
+}
+
+func (d *destructKeyedSource) MemUsage(ctx *MemUsageContext) (uint64, error) {
```
Comment (Gabri3l) [inline destruct.go:264]: Added the expected `MemUsage` function

**Review by tsedgwick — APPROVED** [review]
LGTM

File: destruct.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	return d.wrapped.MemUsage(ctx)
```
Comment (arahmanan) [inline destruct.go:265]: [opt] Should we check if `wrapped` is nil beforehand?

**Review by arahmanan — APPROVED** [review]
LGTM! just an optional comment from my end.

**Pull request merged**